File: 2021/Day3/Valid parantheses/Valid_parantheses.cpp
class Solution
{
public:
    bool isValid(string s)
    {
        std::unordered_map<char, char> match{{')', '('}, {']', '['}, {'}', '{'}};
        std::deque<char> stack;
        for (const auto &c : s)
        {
            if (c == '(' || c == '[' || c == '{')
                stack.push_back(c);
            else
            {
                if (stack.empty() || stack.back() != match[c])
                    return false;
                stack.pop_back();
            }
        }
        return (stack.empty());
    }
};

File: 2021/Day3/Valid parantheses/Valid_parantheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) < 2:
            return False

        parentHash = {"(": ")", "{": "}", "[": "]"}
        stack = []

        for i in range(len(s)):
            if s[i] == "(" or s[i] == "{" or s[i] == "[":
                stack.append(s[i])
            else:
                if len(stack) < 1:
                    return False

                if not s[i] == parentHash[stack.pop(-1)]:
                    return False

        return len(stack) == 0
